### DISCUSSION TITLE
Prime Tree: improved filtering

### DISCUSSION THREAD

**Discussion by ray007** [primefaces/community#1645] [PrimeVue]
Filtering of trees is currently very simple: a `bool` config option `filter`, which shows an additional text input and does only basic filtering by text.

I'm trying to upgrade to PrimeVue from my old ui.
There I have a dropdown beside the searchbox, where I can select node types.
So filtering is done by text (label) and type (from list of checkboxes).

Unfortunately, there's no slot to design the filter input(s), but doing my own ui for that is easy.

But there's currently no way to supply a filter function and trigger filtering from outside - can we get something for that?

**Comment by erikherberg** [comment]
I completely agree that the current filtering implementation in PrimeVue’s Tree component is quite limited. The ability to provide a custom filtering function and trigger filtering externally would be a great improvement.  

### My Use Case

In my case, filtering is done remotely, meaning I don’t want to use PrimeVue’s internal filtering logic at all. Instead, I’d like to:  
1. Use the built-in UI filter elements but override the filtering logic entirely.  
2. Have the ability to trigger filtering externally—this would be useful for cases like debounced search, server-side filtering, or when filters are more complex (e.g., filtering by multiple attributes).  
3. Extend this capability to **TreeSelect**

### Suggested Improvements  

- **Custom Filtering Function**: Allow users to provide a filtering function that takes the current filter value and returns the filtered nodes.  
- **TreeSelect Filtering Support**: Expose an event (e.g., `@filter`) so that developers can listen for filter changes and trigger remote filtering accordingly.  

Would love to hear thoughts on whether this is something that could be or if there are workarounds available.
